File: RobWork/example/snippets/ex-print-kinematic-tree.cpp
#include <rw/kinematics/Frame.hpp>
#include <rw/models/WorkCell.hpp>

#include <string>

using namespace rw::kinematics;
using rw::models::WorkCell;
using rw::math::Transform3D;

void printKinematicTree(
    Frame& frame,
    const State& state,
    const Transform3D<>& parentTransform,
    int level)
{
    const Transform3D<> transform = parentTransform * frame.getTransform(state);

    std::cout
        << std::string(level, ' ')
        << frame.getName()
        << " at "
        << transform.P()
        << std::endl;

    for(Frame::Ptr& child: frame.getChildrenList(state)) {
        printKinematicTree(*child.get(), state, transform, level + 1);
    }

    /*for(Frame& child: frame.getChildrenList(state)) {
        printKinematicTree(child, state, transform, level + 1);
    }*/
}

void printDefaultWorkCellStructure(const WorkCell& workcell)
{
    std::cout << workcell << std::endl;
    printKinematicTree(
        *workcell.getWorldFrame(),
        workcell.getDefaultState(),
        Transform3D<>::identity(),
        0);
}

File: RobWork/example/java/src/ExPrintKinematicTree.java
import org.robwork.sdurw.*;

public class ExPrintKinematicTree {
    public static void printKinematicTree(
            Frame frame,
            State state,
            Transform3Dd parentTransform,
            int level)
    {
        final Transform3Dd transform = parentTransform.multiply(frame.getTransform(state));

        for (int i = 0; i < level; i++) {
            System.out.print(" ");
        }
        System.out.println(frame.getName() + " at " + transform.P());

        FrameVector children = frame.getChildren(state);
        for (int i = 0; i < children.size(); i++) {
            printKinematicTree(children.get(i), state, transform, level + 1);
        }
    }

    public static void printDefaultWorkCellStructure(WorkCellPtr workcell)
    {
        printKinematicTree(
                workcell.getWorldFrame(),
                workcell.getDefaultState(),
                Transform3Dd.identity(),
                0);
    }

}
